animation-glib/constructor-helpers.cpp

```diff
@@ -0,0 +1,156 @@
+/**
+ * replace_construct_param:
+ * @construct_params: (array length=n_construct_params) An array of #GObjectConstructParam
+ * @n_construct_params: Number of elements in @construct_params.
+ * @prop_name: The name of the construct prop to replace the value of.
+ * @initialize_func: A function which sets the GValue to something sensible.
+ *
+ * Replace a construction parameter @prop_name in the
+ * passed @construct_params by using the passed @initialize_func.
+ *
+ * This function must always replace one construct parameter, it is
+ * an error to pass a @prop_name that is not in the @construct_params.
+ */
+void
+replace_construct_param (GObjectConstructParam                          *construct_params,
+                         unsigned int                                    n_construct_params,
+                         const char                                     *prop_name,
+                         AnimationConstructorHelpersInitializeValueFunc  initialize_func,
+                         gpointer                                        initialize_func_data)
+{
+  /* The prop should always be found in the array so that we can replace
+   * it, this function doesn't support appending the prop. That means
+   * that the relevant prop must always G_PARAM_CONSTRUCT or
+   * G_PARAM_CONSTRUCT_ONLY. */
+  for (unsigned int i = 0; i < n_construct_params; ++i)
+    {
+      if (g_strcmp0 (construct_params[i].pspec->name, prop_name) == 0)
+        {
+          g_value_unset (construct_params[i].value);
+          initialize_func (construct_params[i].value, initialize_func_data);
+          return;
+        }
+    }
+
+  g_assert_not_reached ();
+}
+
+template <typename FunctionType, typename ...Args>
+static typename std::result_of <FunctionType(Args...)>::type
+invoke_function_thunk (Args... args, gpointer lambda)
+{
+  FunctionType *f = reinterpret_cast <FunctionType *> (lambda);
+
+  return (*f)(args...);
+}
+
+
+void
+replace_named_pointer_prop_in_construct_params (GObjectConstructParam *construct_params,
+                                                unsigned int           n_construct_params,
+                                                const char            *prop_name,
+                                                gpointer               ptr)
+{
+  auto set_value = [ptr](GValue *value) {
+    g_value_init (value, G_TYPE_POINTER);
+    g_value_set_pointer (value, ptr);
+  };
+  replace_construct_param (construct_params,
+                           n_construct_params,
+                           prop_name,
+                           (AnimationConstructorHelpersInitializeValueFunc) invoke_function_thunk <decltype (set_value), GValue *>,
+                           &set_value);
+}
+
+void
+replace_interface_prop_in_construct_params (GObjectConstructParam *construct_params,
+                                            unsigned int           n_construct_params,
+                                            gpointer               interface)
```

Comment:
I mentioned this earlier but I'm not a big fan of encoding the interface as a gobject property. Why is constructing the interface after the gobject constructor ran not an option? The `constructed` GObject vfunc exists already that would guarantee those semantics.

Reply:
Hmm, I suppose what I wanted to do here was construct the underlying interface as soon as we had all the construct properties without having to keep them around. I remember that in the wobbly model gobject binding using the constructor was the best way to do this.

I was thinking earlier that we could jettison the interface prop and just set an internal member but now that I think of it, such an approach wont work because we dont have access to the private data yet :-(